File: Day3/day3.R
library(tidyverse)

input <- readLines("Day3/input.txt")

# part1 ####
get_common <- function(sack) {
  map_chr(sack, function(sack) {
    split <- str_split(sack,"", simplify = T)
    c1 <- split[1:length(split)/2]
    c2 <- split[(length(split)/2+1):length(split)]
    common <- intersect(c1,c2)
    common
  })
}

scoring <- c(setNames(1:26,letters),setNames(27:52,LETTERS))

get_score <- function(letter) {
  recode(letter, !!!scoring)
}

get_common(input) %>%
  get_score() %>%
  sum()

# part2 ####
groups_f <- (seq_along(input)-1)%/%3
groups <- split(input, groups_f)

intersect_sack <- function(sack1,sack2) {
  s1 <- str_split(sack1,"", simplify = T)
  s2 <- str_split(sack2,"", simplify = T)
  intersect(s1,s2)
}
get_badge <- function(group) {
  group %>%
    purrr::reduce(intersect_sack)
}

map_chr(groups,get_badge) %>%
  get_score() %>%
  sum()

File: Day3/day3.py
with open('Day3/input.txt') as file:
     lines = [line.rstrip() for line in file]

# PART 1

def get_common(sack): 
  mid = int(len(sack)/2)
  c1 = sack[0:mid]
  c2 =sack[mid:]
  common = set(c1).intersection(c2)
  return "".join(common)

commons = [*map(get_common, lines)]

import string
import numpy as np

scores = string.ascii_lowercase+string.ascii_uppercase

def get_score(letter):
  score = scores.find(letter)+1
  return score

sum([*map(get_score, commons)])

# PART2

groups_f = [*range(0,int(len(lines)),3)]
groups = [lines[i:(i+3)] for i in groups_f]

def get_common2(group):
  common = set(group[0]).intersection(group[1]).intersection(group[2])
  return "".join(common)

commons2 = [get_common2(group) for group in groups]

sum([*map(get_score, commons2)])
